client/state/selectors/is-eligible-for-upwork-support.ts

```diff
@@ -12,22 +12,22 @@ import { getCurrentUserLocale } from 'state/current-user/selectors';
 import getSitesItems from 'state/selectors/get-sites-items';
 import { isBusinessPlan, isEcommercePlan } from 'lib/plans';
 
-const SPANISH_LANG_SLUGS = [ 'es', 'es-cl', 'es-mx' ];
+const UPWORK_LOCALES = [ 'es', 'es-cl', 'es-mx', 'fr', 'fr-ca', 'fr-be', 'fr-ch' ];
 
 /**
  * @param {Object} state Global state tree
  * @return {Boolean} Whether or not this customer should receive Upwork support
  */
-export default function isEligibleForUpworkSupport( state ) {
-	// Upwork is currently available for Spanish-speaking users only
-	if ( ! includes( SPANISH_LANG_SLUGS, getCurrentUserLocale( state ) ) ) {
+export default function isEligibleForUpworkSupport( state ): boolean {
```

Comment:
did it fail to detect `boolean` automatically?

the only thing I see ambiguous here is `isBusinessPlan()` and `isEcommercePlan()` and those trace out to boolean as well, though they _can_ throw errors.

Reply:
I'm not sure in lieu of the above. Do you want me to remove?

It maybe worth considering explicitly typing _exported_ functions as much as you can. In the Happy Tools codebase I've seen benefit in explicitly typing exports. It has helped the checker in more than one situation and (in theory) Flow's strength is its inference.

Just my two cents. @beaucollins may have thoughts to add one direction or anything here.

Reply:
it wouldn't take much to persuade me of what you're saying. I like explicitly typing interfaces (exported functions). I'm in that area where I just want to asks and check our assumptions. does the additional `boolean` help more people or confuse people? I don't know. I only asked in this case because it seemed like the inference is strong and exclusively `boolean`, so my working principle is something like "if TypeScript finds the right types automatically then let it be."

I claim very little confidence in that or any approach at this stage of the rollout.

Reply:
> "if TypeScript finds the right types automatically then let it be."

I had this opinion of Flow too until using it more and more. TypeScript may be better but I found when the type checker starts giving you type errors, tracking down implicitly typed problems usually is less clear for the future developer than explicitly typed.

If I as a developer am able to discern that a function should always return a specific type I see no reason not to define it. The type checker will enforce the rules for you. I find it even helps me design my interfaces better: start with the types.

With Happy Tools I find it always more pleasant to have explicit types when type failures happen.

> does the additional boolean help more people or confuse people

Can you give me an example where adding a return type would make something more confusing?

For example, I would assume adding a JSDoc `@returns` attributes to comment blocks would _always_ be considered good form. This is the same thing as adding a `@returns` but also at the same time allowing a computer to check that it is correct.

While we were adding JSDoc attributes to our source code we were going through the motions of adding explicit types but never getting the benefits of it. We practiced the incantations but never got to see any of the magic.

Reply:
> For example, I would assume adding a JSDoc @returns attribute to comment blocks would always be considered good form. This is the same thing as adding a @returns but also at the same time allowing a computer to check that the function is correct instead of hoping that the comment is correct.

@beaucollins I agree entirely but in conversations with people I've repeatedly heard that the type annotations are more confusing. I can't explain it other than the fact that it's a different thing or syntax is more scary than it needs to be.

> I had this opinion of Flow too until using it more and more…tracking down implicitly typed problems usually is less clear for the future developer than explicitly typed.

It's encouraging to hear this experience. Can you share an example of something that helped persuade you to change, or was it merely that you couldn't trace the types back to the function providing the constraint?

Since I know you enjoy types yourself have you gotten similar sentiments from developers you work with that weren't already acquainted with and excited about Flow?

Reply:
> Since I know you enjoy types yourself have you gotten similar sentiments from developers you work with that weren't already acquainted with and excited about Flow?

Yes! I was initially opposed (of course, to any change) but once we started flowing the codebase, it began to help me make less mistakes. Many times, I was about to commit a piece of code and then flow stopped me. I was like: "why? what's wrong? this is a normal looking js code!". But after a deeper inspection, flow revealed there could be a situation in which that code would error out or simply not work as intended.

So IMO, it's worth it.

Now, whether flow vs typescript: haven't used typescript but folks around me argue it's better than flow. I guess they haven't used flow 😅 Since calypso (and I guess broader a8c) has agreed to use TS, we might rethink our flow usage later but for now, it's not worth refactoring it all.

Reply:
> I agree entirely but in conversations with people I've repeatedly heard that the type annotations are more confusing. I can't explain it other than the fact that it's a different thing or syntax is more scary than it needs to be.

I'm sympathetic to this. I jumped into Flow after it was introduced to our codebase _and_ I was and remain concerned about "syntax soup" and readability concerns.

I don't have easy answers and I realize a project as big as Calypso makes wider turns. For me, personally, getting to the point where we intentionally _removed_ JSDoc annotations _really_ helped. The "double annotations" were a mental/readability mess for me and when all that "noise" went to one human+computer friendly approach it ended up clicking together nicely.

Reply:
thanks for sharing. getting rid of JSDoc comments and `PropTypes` is something I consider important as well.

for related issues I'm entirely opposed to using `any` in this hybrid codebase because I don't want us to hide type errors that we could later come back and fix. it's harder for me to understand given that Calypso is full of type errors all over the place and I recognize we can't fix it all in any reasonable frame of time.

maybe it's reasonable to say we add annotations when we're working in a function and we have a reasonable confidence that the type is the type we want.

Reply:
> It's encouraging to hear this experience. Can you share an example of something that helped persuade you to change, or was it merely that you couldn't trace the types back to the function providing the constraint?

https://github.com/Automattic/happy-tools/pull/2455#pullrequestreview-228885161

The comment linked to above is an example where implicit types resulted in the incorrect initial fix, even though it correctly fixed the type issue. Flow identified that there was a type issue, but the real problem didn't become immediately obvious until I added an explicit type to

```js
const renderComponentInLayout = (context, component: React.Node) => {
```

The final fix does not include the added type (`component: React.Node`) but having explicit types makes Flow's type errors more obvious. So then the question becomes why remove the types if they make future problems easier to solve?

Types _are_ documentation. If a developer can bring themselves to understand JSDoc syntax I see no reason why the TS types are somehow a bigger hurdle to understanding besides resistance to "new".

Reply:
@dmsnell I need to get this merged. What's the verdict here on the boolean typing? Your call.

Reply:
sorry - go for it @hewsut - I should have gave you the go-ahead to not let this be a blocker. I'm grateful for y'all sharing your input and experience

Reply:
No worries. We put it in early and had time to spare. Just coming due.

Reply:
@beaucollins you make a pretty compelling point. one thing that has been bothering me in reviewing the typed PRs the past week has been an aspect of how we're not constraining ourselves to fix type errors.

for Calypso I think this is imperative if we're going to introduce them at all because we just don't have the bandwidth to fix every problem in a file if we want to add a few types.

when I read the reasoning here for explicit types I hear the kind of reasoning I too affirm:
 - we have to deal with types now or later
 - slowing down to define an interface is the faster option
 - explicit types prevent bugs

a way that bothers me about this gradual rollout where we're not using the TypeScript binary to compile (or Flow to type-check) is that we won't be alerted that our change-of-type to a function will break calling functions. we still have to manually jump to those files that rely on this function and then see if an error appears there. `tsc` would make these connections for us and tell us of the problem, but I fear right now we'll only see the errors surface at the call site (or as one type error among 100,000 if we run `tsc` manually).

no solution now, just questions.

![DisconnectedCallSites](https://user-images.githubusercontent.com/5431237/57039987-c7029080-6c23-11e9-99e8-26954b4fc93e.gif)